src/main/java/com/atlassian/bitbucket/jenkins/internal/client/ModernBitbucketBuildStatusClientImpl.java

```diff
@@ -1,26 +1,61 @@
+import com.atlassian.bitbucket.jenkins.internal.provider.DefaultInstanceIdentityProvider;
+import com.atlassian.bitbucket.jenkins.internal.provider.InstanceIdentityProvider;
+import com.google.common.annotations.VisibleForTesting;
 import okhttp3.HttpUrl;
 
+import java.security.Signature;
+import java.security.interfaces.RSAPrivateKey;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+
 import static java.util.Objects.requireNonNull;
 import static org.apache.commons.lang3.StringUtils.stripToNull;
 
 public class ModernBitbucketBuildStatusClientImpl implements BitbucketBuildStatusClient {
 
+    private static final String BBS_BUILD_STATUS_SIGNATURE_ID = "BBS-Signature";
     private static final String BUILD_STATUS_VERSION = "1.0";
     private final BitbucketRequestExecutor bitbucketRequestExecutor;
-    private final String revisionSha;
+    private final InstanceIdentityProvider instanceIdentityProvider;
     private final String projectKey;
     private final String repoSlug;
+    private final String revisionSha;
 
-    ModernBitbucketBuildStatusClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor, String projectKey, String repoSlug, String revisionSha) {
+    @VisibleForTesting
+    ModernBitbucketBuildStatusClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor, String projectKey,
+                                         String repoSlug, String revisionSha,
+                                         InstanceIdentityProvider instanceIdentityProvider) {
         this.bitbucketRequestExecutor = requireNonNull(bitbucketRequestExecutor, "bitbucketRequestExecutor");
+        this.instanceIdentityProvider = requireNonNull(instanceIdentityProvider, "instanceIdentityProvider");
         this.revisionSha = requireNonNull(stripToNull(revisionSha), "revisionSha");
         this.projectKey = requireNonNull(stripToNull(projectKey), "projectKey");
         this.repoSlug = requireNonNull(stripToNull(repoSlug), "repoSlug");
     }
 
+    ModernBitbucketBuildStatusClientImpl(BitbucketRequestExecutor bitbucketRequestExecutor, String projectKey,
+                                         String repoSlug, String revisionSha) {
+        this(bitbucketRequestExecutor, projectKey, repoSlug, revisionSha, new DefaultInstanceIdentityProvider());
+    }
+
+    public Map<String, String> generateSignature(BitbucketBuildStatus buildStatus) {
+        Map<String, String> headers = new HashMap<>();
+        try {
+            RSAPrivateKey key = instanceIdentityProvider.getInstanceIdentity().getPrivate();
+            Signature sig = Signature.getInstance("SHA256with" + key.getAlgorithm());
+            sig.initSign(key);
+            sig.update(buildStatus.getSignature().getBytes("UTF-8"));
+            headers.put(BBS_BUILD_STATUS_SIGNATURE_ID, Base64.getEncoder().encodeToString(sig.sign()));
+        } catch (Exception e) {
+            throw new BitbucketClientException("Exception signing build status", e);
```

Comment:
Log+throw is generally considered an [antipattern](https://community.oracle.com/docs/DOC-983543#antipatterns) so I think we should not do that. It usually leads to the exception being logged twice which is annoying.

Reply:
Good to know. I generally have used this (anti-)pattern when we log significantly more details than we put in the exceptions, which is especially the case in Cloud services, where you want to log some potentially revealing details (of course with all the relevant considerations for privacy, etc.), but you don't include those details in the exceptions in case they leak to the UI...
Anyways, in this case let's follow whatever the right pattern is. :+1:

Reply:
I think in this case, the exception will bubble up and produce the right error message, so I don't think we need extra logging. I can see the log -> throw pattern making sense if you expect that exception to get munched, which it does sometimes in Jenkins, but in this case I think we're good.